client/rest-high-level/src/main/java/org/elasticsearch/client/security/user/privileges/Role.java

```diff
@@ -0,0 +1,271 @@
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.optionalConstructorArg;
+
+public final class Role implements ToXContentObject {
+
+    public static final ParseField CLUSTER = new ParseField("cluster");
+    public static final ParseField GLOBAL = new ParseField("global");
+    public static final ParseField INDICES = new ParseField("indices");
+    public static final ParseField APPLICATIONS = new ParseField("applications");
+    public static final ParseField RUN_AS = new ParseField("run_as");
+    public static final ParseField METADATA = new ParseField("metadata");
+
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<Role, Void> PARSER = new ConstructingObjectParser<>("role_descriptor", false,
+            constructorObjects -> {
+                int i = 0;
+                final Collection<ClusterPrivilege> clusterPrivileges = (Collection<ClusterPrivilege>) constructorObjects[i++];
+                final ManageApplicationsPrivilege manageApplicationPrivileges = (ManageApplicationsPrivilege) constructorObjects[i++];
+                final Collection<IndicesPrivileges> indicesPrivileges = (Collection<IndicesPrivileges>) constructorObjects[i++];
+                final Collection<ApplicationResourcePrivileges> applicationResourcePrivileges =
+                        (Collection<ApplicationResourcePrivileges>) constructorObjects[i++];
+                final Collection<String> runAs = (Collection<String>) constructorObjects[i++];
+                final Map<String, Object> metadata = (Map<String, Object>) constructorObjects[i];
+                return new Role(clusterPrivileges, manageApplicationPrivileges, indicesPrivileges, applicationResourcePrivileges,
+                        runAs, metadata);
+            });
+
+    static {
+        PARSER.declareFieldArray(optionalConstructorArg(), (p, c) -> ClusterPrivilege.fromString(p.text()), CLUSTER,
+                ValueType.STRING_ARRAY);
+        PARSER.declareObject(optionalConstructorArg(), ManageApplicationsPrivilege.PARSER, GLOBAL);
+        PARSER.declareFieldArray(optionalConstructorArg(), IndicesPrivileges.PARSER, INDICES, ValueType.OBJECT_ARRAY);
+        PARSER.declareFieldArray(optionalConstructorArg(), ApplicationResourcePrivileges.PARSER, APPLICATIONS, ValueType.OBJECT_ARRAY);
+        PARSER.declareStringArray(optionalConstructorArg(), RUN_AS);
+        PARSER.<Map<String, Object>>declareObject(constructorArg(), (parser, c) -> parser.map(), METADATA);
+    }
+
+    private final Set<ClusterPrivilege> clusterPrivileges;
+    private final @Nullable ManageApplicationsPrivilege manageApplicationPrivileges;
```

Comment:
In my view when we will be adding another type  of "global" privilege, there will simply be another privilege in this `Role` definition on the client side. Yes, there will be forward compatibility issues (some new client added a new privilege that an older client cannot discern) but how can we account for this. 

What does the `ConditionalClusterPrivileges` (the base class of `ManageApplicationsPrivilege` on the server side) has, that we need to model on the client side. Even if we would be modeling a base class for `ManageApplicationsPrivilege`, how would that be helpful?

> It really needs to be some sort of abstraction of "global" privileges so that we can add more in the future without unnecessarily breaking the's client API.

Again, the answer we all have in our head is to parse the "global" privilege as a `Map`. It has forward compatibility (in the sense that we don't fail to deserialize) but it is less helpful even for clients that stay on version par.

Reply:
`ManageApplicationsPrivilege` is gone and has been replaced with `GlobalPrivileges` .
`GlobalPrivileges` comprises of nested named objects (under the `global.application` namespace). Each such object is a privilege. There is no restriction on the names (they are strings). This way we have the extensibility you argue for, but the client loses the ability to define `manage application privilege` in a simple way (he has to know the `manage` name for the privilege and that the privilege has a single `applications` key with the list value of the application names he is allowed to manage).